add a checkbox for the schedule report pop up

Ruiqian1 opened the pull request:
Resolves #

<!-- The above line will close the issue upon merge -->

## We want to add the functionality of adding an option that lets users to choose if they want the report to be sent even if the report is empty

## We add a checkbox at the schedule report pop up

-

## 
![Screen Shot 2019-07-17 at 9 43 47 AM](https://user-images.githubusercontent.com/10789893/61385075-688a9d00-a877-11e9-92cb-0655518f04e4.png)


## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

cythrawll commented:
Nice, we have been thinking about doing something similar.

* Add this to the`schedulingRules` hash in the `deliveryRule` model so it gets persisted correctly.
* Put behind a featureFlag so we can hide this option until we have support in the jobs-service for this.

balajimanoharan commented:
We had an initial design for this feature. The checkboxes would be rendered below the dropdowns. (screenshot below). Can we do something like that?

![Screen Shot 2019-07-17 at 10 59 49 AM](https://user-images.githubusercontent.com/2703389/61391375-316db900-a882-11e9-9ff1-c3f1130adc3f.png)

Ruiqian1 commented:
@balajimanoharan @cythrawll I have updated the schedule report model and put the checkbox behind a feature flag. The layout has also been adjusted. Please review and let me know if there are any further questions.
<img width="686" alt="Screen Shot 2019-07-18 at 1 41 17 PM" src="https://user-images.githubusercontent.com/10789893/61483424-5fc2c580-a962-11e9-95d2-ae7f14eef4e9.png">

mattgraham1995 commented on packages/core/addon/models/delivery-rule.js:
```diff
@@ -37,6 +37,7 @@ export default DS.Model.extend(Validations, {
   format: DS.attr({ defaultValue: () => {} }),
   recipients: DS.attr({ defaultValue: () => [] }),
   version: DS.attr('number', { defaultValue: '1' }),
+  notifyWhenEmpty: DS.attr('number', { defaultValue: '0' }),
```
can we use `boolean` instead of `number`?

cythrawll replied:
Actually this needs to be in scheduling rules, not it's own property.

Ruiqian1 replied:
Can you explaining a little more on why we need to put this under schedulingRules instead of at the same level of format, frequency, and recipients?

jkusa commented:
Can we use a toggle instead of checkbox?

jkusa commented:
Can we use a name that indicates that the system should not send a notification when there is an empty result?

Ruiqian1 commented:
@jkusa I have changed the checkbox to be a toggle and revised the label. Here is how is looks like now
<img width="730" alt="Screen Shot 2019-07-19 at 3 43 21 PM" src="https://user-images.githubusercontent.com/10789893/61564564-560d9080-aa3c-11e9-8f01-8881a0d57e5a.png">

Ruiqian1 commented:
@cythrawll In the schedule rule UI, we have frequency, format, as well this boolean flag value indicating whether we want report to be sent if it is empty. I think all those 3 options should be of the equivalent to each other. For this reason, I think it would be find to have this boolean value as a separate attribute just as frequency and format.

jkusa commented on packages/reports/addon/templates/components/common-actions/schedule.hbs:
```diff
@@ -81,6 +81,17 @@
+      {{#if (feature-flag "enableNotifyWhenEmpty")}}
+        <div class="schedule-modal__section-group">
+          {{x-toggle
+            theme="ios"
```
Can we remove this to match the other toggles

jkusa replied:
Can we add some margin around this component it seems a little close to the select above

Ruiqian1 replied:
cool. I am not an apple fan tbh XD

jkusa commented on packages/reports/addon/templates/components/common-actions/schedule.hbs:
```diff
@@ -81,6 +81,17 @@
+            class="schedule-modal__toggle schedule-modal__toggle--notifyWhenEmpty"
+            value=localDeliveryRule.notifyWhenEmpty
+            onToggle=(mut localDeliveryRule.notifyWhenEmpty)}}
+          <div class="schedule-modal__label">Do Not Send The Report If It Is Empty</div>
```
`Only send if data is present`?

Ruiqian1 replied:
sure

jkusa commented on packages/core/addon/models/delivery-rule.js:
```diff
@@ -37,6 +37,7 @@ export default DS.Model.extend(Validations, {
   format: DS.attr({ defaultValue: () => {} }),
   recipients: DS.attr({ defaultValue: () => [] }),
   version: DS.attr('number', { defaultValue: '1' }),
+  notifyWhenEmpty: DS.attr('boolean', { defaultValue: false }),
```
Can we move this to the `schedulingRules` and call it `mustHaveData` with a default of `false`

`schedulingRules` is intended to be where we put advanced settings.

jkusa commented on packages/reports/tests/dummy/config/environment.js:
```diff
@@ -42,6 +42,7 @@ module.exports = function(environment) {
       FEATURES: {
         enableScheduleReports: true,
         enableMultipleExport: true,
+        enableNotifyWhenEmpty: true,
```
`enabledNotifyIfData`

Ruiqian1 commented:
@jkusa I have just changed the variable name from `notifyWhenEmpty` to `mustHaveData`, I also added a 5px margin at the top of the div that contains the toggle. Please let me know if you have further concerns.

Ruiqian1 commented:
@jkusa @cythrawll @balajimanoharan Hi, I have addressed the comments. Please take a look of the PR when you have time. Thanks :)

mattgraham1995 commented:
In the `deliveryRule` model, could we set the default value for `mustHaveData` in `schedulingRules`?

`schedulingRules: DS.attr({ defaultValue: () => {} })` 
=>
`schedulingRules: DS.attr({ defaultValue: () => { 
return featureFlag('enabledNotifyIfData') ? { mustHaveData: false } : {}; })`

balajimanoharan commented on packages/reports/addon/templates/components/common-actions/schedule.hbs:
```diff
@@ -81,6 +81,16 @@
+        <div class="schedule-modal__section-group">
+          {{x-toggle
+            size="small"
+            class="schedule-modal__toggle schedule-modal__toggle--notifyWhenEmpty"
```
Can we rename the class name to `schedule-modal__must-have-data-toggle`? 
We don't use camel case for class names.

Ruiqian1 replied:
The convention we have for class name is schedule-modal__{type}--{name}. I will change it to `schedule-modal__toggle--must-have-data`

balajimanoharan replied:
actually the convention we follow is called BEM. You can check this out for what BEM stands for.
https://css-tricks.com/bem-101/

Ruiqian1 replied:
Oh. That's awesome. Thanks!

balajimanoharan replied:
`must-have-data` is not a modifier. Modifiers are things like `disabled`, `active`, `invalid`, `red`, etc.. Which is why `schedule-modal__must-have-data-toggle` is a better name according to the convention.

Ruiqian1 replied:
I see. Let me change that!

Ruiqian1 replied:
Just addressed the naming convention concern.

balajimanoharan commented on packages/reports/tests/acceptance/schedule-modal-test.js:
```diff
@@ -93,6 +97,9 @@ module('Acceptance | Navi Report Schedule Modal', function(hooks) {
     await click('.schedule-modal__dropdown--frequency .ember-power-select-trigger');
     await click($('.ember-power-select-option:contains(Day)')[0]);
 
+    // Set notify when empty to be true
```
I think we can remove this comment.

balajimanoharan commented on packages/reports/tests/acceptance/schedule-modal-test.js:
```diff
@@ -117,6 +124,10 @@ module('Acceptance | Navi Report Schedule Modal', function(hooks) {
 
+    assert
+      .dom('.schedule-modal__must-have-data-toggle .x-toggle')
+      .isChecked('the notify when empty toggle should be toggled on after toggling it on');
```
`must have data toggle is checked when clicked`

balajimanoharan commented on packages/reports/tests/acceptance/schedule-modal-test.js:
```diff
@@ -210,6 +221,9 @@ module('Acceptance | Navi Report Schedule Modal', function(hooks) {
     await click('.schedule-modal__dropdown--frequency .ember-power-select-trigger');
     await click($('.ember-power-select-option:contains(Day)')[0]);
 
+    // Set notify when empty to be true
```
remove comment, we don't have a `notifyWhenEmpty` property anymore.

balajimanoharan commented on packages/reports/tests/acceptance/schedule-modal-test.js:
```diff
@@ -244,6 +262,9 @@ module('Acceptance | Navi Report Schedule Modal', function(hooks) {
     await click('.schedule-modal__dropdown--frequency .ember-power-select-trigger');
     await click($('.ember-power-select-option:contains(Day)')[0]);
 
+    // Set notify when empty to be true
```
remove comment.

mattgraham1995 replied:
Are we going to update these comments? The component is called `must have data` instead of `notify when empty` now so the comments don't really make sense

balajimanoharan commented on packages/reports/tests/acceptance/schedule-modal-test.js:
```diff
@@ -261,6 +282,10 @@ module('Acceptance | Navi Report Schedule Modal', function(hooks) {
 
+    assert
+      .dom('.schedule-modal__must-have-data-toggle .x-toggle')
+      .isNotDisabled('notify when empty changes to an existing schedule are discarded after clicking cancel');
```
Did you mean to check `isNotChecked` here instead of `isNotDisabled`?

Ruiqian1 replied:
Good catch! Just changed it.

Ruiqian1 commented:
@cythrawll Thank you for reviewing this PR. I don't see to have access to merge it. Could you please grant me the access if possible or could you pleases merge the PR?
